contrib/rabbitmq/pkg/adapter/adapter.go

```diff
@@ -0,0 +1,262 @@
+	"context"
+	"encoding/json"
+	"github.com/sbcd90/wabbit"
+	"github.com/sbcd90/wabbit/amqp"
+	"github.com/cloudevents/sdk-go/pkg/cloudevents"
+	"github.com/cloudevents/sdk-go/pkg/cloudevents/client"
+	"github.com/cloudevents/sdk-go/pkg/cloudevents/types"
+	"github.com/knative/eventing-sources/pkg/kncloudevents"
+	"github.com/knative/pkg/logging"
+	"go.uber.org/zap"
+	"strings"
+)
+
+const (
+	eventType = "dev.knative.rabbitmq.event"
+)
+
+type ExchangeConfig struct {
+	Name        string
+	TypeOf      string
+	Durable     bool
+	AutoDeleted bool
+	Internal    bool
+	NoWait      bool
+}
+
+type ChannelConfig struct {
+	PrefetchCount int
+	GlobalQos     bool
+}
+
+type QueueConfig struct {
+	Name             string
+	RoutingKey       string
+	Durable          bool
+	DeleteWhenUnused bool
+	Exclusive        bool
+	NoWait           bool
+}
+
+type Adapter struct {
+	Brokers        string
+	Topic          string
+	ChannelConfig  ChannelConfig
+	ExchangeConfig ExchangeConfig
+	QueueConfig    QueueConfig
+	SinkURI        string
+	client         client.Client
+}
+
+func (a *Adapter) initClient() error {
+	if a.client == nil {
+		var err error
+		if a.client, err = kncloudevents.NewDefaultClient(a.SinkURI); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (a *Adapter) Start(ctx context.Context, stopCh <-chan struct{}) error {
+	logger := logging.FromContext(ctx)
+
+	if err := a.initClient(); err != nil {
+		logger.Error("Failed to create cloudevent client", zap.Error(err))
+		return err
+	}
+
+	logger.Info("Starting with config: ", zap.Any("adapter", a))
+
+	conn, err := amqp.Dial(a.Brokers)
+	if err != nil {
+		logger.Error(err)
+	}
+	defer conn.Close()
+
+	ch, err := conn.Channel()
+	if err != nil {
+		logger.Error(err)
+	}
+	defer ch.Close()
+
+	err = ch.Qos(a.ChannelConfig.PrefetchCount,
+		0,
+		a.ChannelConfig.GlobalQos)
+	if err != nil {
+		logger.Error(err)
+	}
+
+	queue, err := a.StartAmqpClient(ctx, &ch)
+	if err != nil {
+		logger.Error(err)
+	}
+
+	return a.pollForMessages(ctx, &ch, queue, stopCh)
+}
+
+func (a *Adapter) StartAmqpClient(ctx context.Context, ch *wabbit.Channel) (*wabbit.Queue, error) {
+	logger := logging.FromContext(ctx)
+	exchangeConfig := fillDefaultValuesForExchangeConfig(&a.ExchangeConfig, a.Topic)
+
+	err := (*ch).ExchangeDeclare(
+		exchangeConfig.Name,
+		exchangeConfig.TypeOf,
+		wabbit.Option{
+			"durable":  exchangeConfig.Durable,
+			"delete":   exchangeConfig.AutoDeleted,
+			"internal": exchangeConfig.Internal,
+			"noWait":   exchangeConfig.NoWait,
+		},)
+	if err != nil {
+		logger.Error(err)
+		return nil, err
+	}
+
+	queue, err := (*ch).QueueDeclare(
+		a.QueueConfig.Name,
+		wabbit.Option{
+			"durable":   a.QueueConfig.Durable,
+			"delete":    a.QueueConfig.DeleteWhenUnused,
+			"exclusive": a.QueueConfig.Exclusive,
+			"noWait":    a.QueueConfig.NoWait,
+		},)
+	if err != nil {
+		logger.Error(err)
+		return nil, err
+	}
+
+	routingKeys := strings.Split(a.QueueConfig.RoutingKey, ",")
```

Comment:
Keep in mind that `fanout` exchanges don't use `routingKeys`. In this case your code will leave the queue without a binding and won't receive messages.

Reply:
Hi @leoluz, thanks for your comment. I'll work on this.

Reply:
Hi @leoluz , I have fixed this issue now.